File: 2018/09/Marbles.java

public class Marbles {
  public static class Node {
    private long value;
    private Node prev;
    private Node next;
    public Node(long value, Node prev, Node next) {
      this.value = value;
      this.prev = prev;
      this.next = next;
    }

    public void setNext(Node next) {
      this.next = next;
    }

    public void setPrev(Node prev) {
      this.prev = prev;
    }

    public long getValue() {
      return value;
    }

    public Node rotate(int offset) {
      if (offset == 0) {
        return this;
      } else if (offset > 0) {
        return this.next.rotate(offset - 1);
      } else {
        return this.prev.rotate(offset + 1);
      }
    }

    public Node insert(long newValue) {
      Node newNode = new Node(newValue, this, this.next);
      this.next.setPrev(newNode);
      this.setNext(newNode);
      return newNode;
    }

    public long remove() {
      this.next.setPrev(this.prev);
      this.prev.setNext(this.next);
      return this.value;
    }

    public void dump() {
      System.out.print(this.value + " ");
      Node node = this.next;
      while (node.value != this.value) {
        System.out.print(node.value + " ");
        node = node.next;
      }
      System.out.println();
    }
  }

  public static long max(long[] values) {
    long result = values[0];
    for (long value : values) {
      if (value > result) {
        result = value;
      }
    }
    return result;
  }

  public static long[] marbleGame(int numPlayers, int numMarbles) {
    long[] scores = new long[numPlayers];
    Node currentMarble = new Node(0, null, null);
    currentMarble.setNext(currentMarble);
    currentMarble.setPrev(currentMarble);
    for (int i = 1; i < numMarbles; i++) {
      //if (i % 10000 == 0) {
      //  System.out.println(i * 1.0 / numMarbles);
      //}
      int currentPlayer = i % numPlayers;
      if (i % 23 == 0) {
        scores[currentPlayer] += i;
        Node nextMarble = currentMarble.rotate(-6);
        scores[currentPlayer] += currentMarble.rotate(-7).remove();
        currentMarble = nextMarble;
      } else {
        currentMarble = currentMarble.rotate(1).insert(i);
      }
    }
    return scores;
  }

  public static void main(String[] args) {
    System.out.println(max(marbleGame(441, 7103200)));
  }
}

File: 2018/09/marbles.py

def main():
  # Part 1
  #print max(marble_game(10, 1618))
  #print max(marble_game(13, 7999))
  #print max(marble_game(30, 5807))
  #print max(marble_game(441, 71032))
  # Part 2
  #print max(marble_game_2(9, 25))
  #print max(marble_game_2(30, 5807))
  print max(marble_game(441, 7103200))

class Node:
  def __init__(self, value, prev, next):
    self.value = value
    self.prev = prev
    self.next = next

  def rotate(self, offset):
    if offset == 0:
      return self
    elif offset > 0:
      return self.next.rotate(offset - 1)
    else:
      return self.prev.rotate(offset + 1)

  def insert(self, new_value):
    new_node = Node(new_value, self, self.next)
    self.next.prev = new_node
    self.next = new_node
    return new_node

  def remove(self):
    self.next.prev = self.prev
    self.prev.next = self.next
    return self.value

  def dump(self):
    print self.value,
    last = self
    node = self.next
    while node.value != self.value:
      print node.value,
      last = node
      node = node.next
      if node.prev != last:
        raise Exception('bad DLL')
    print

def marble_game_2(num_players, num_marbles):
  scores = [0] * num_players
  current_marble = Node(0, None, None)
  current_marble.next = current_marble
  current_marble.prev = current_marble
  for marble in xrange(1, num_marbles):
    #current_marble.dump()
    #if marble % 10000 == 0:
    #  print marble
    current_player = marble % num_players
    if marble % 23 == 0:
      scores[current_player] += marble
      scores[current_player] += current_marble.rotate(-7).remove()
      current_marble = current_marble.rotate(-6)
    else:
      current_marble = current_marble.rotate(1).insert(marble)
  return scores

def marble_game(num_players, num_marbles):
  scores = [0] * num_players
  current_marble = 0
  marbles = [0]
  for marble in xrange(1, num_marbles):
    if marble % 10000 == 0:
      print marble
    current_player = marble % num_players
    if marble % 23 == 0:
      scores[current_player] += marble
      current_marble = (current_marble - 7 + len(marbles)) % len(marbles)
      scores[current_player] += marbles.pop(current_marble)
    else:
      insert_index = (current_marble + 2) % len(marbles)
      marbles.insert(insert_index, marble)
      current_marble = insert_index
  return scores

if __name__ == '__main__':
  main()
